### PR TITLE
feat: initial feature and setup

### PR DESCRIPTION
**Pull request by daveleek**
## About the changes

The starting point of this plugin. The initial PoC after being converted to typescript and improved, adds unit tests

Documentation PR: #3 

- Releasing is for now done by running `yarn release`, committing `/dist/`, and then tagging
  - Going to automate this with a workflow
- Needs CI workflow
- Going to add tests for the metrics

### Important files

- `src/index.ts` is the entry point for the plugin. It creates and starts the unleash integration that evaluates all configured features and variants
- `src/unleash-action.ts` is the plugin logic that integrates with Unleash through [Unleash Proxy Client JS](https://github.com/Unleash/unleash-proxy-client-js)
- `src/metrics.ts` is a port of `unleash-proxy-client-js` metrics

### Discussion points

- Currently does not implement default values
- Are there more configuration options we should add?
- Does measuring metrics this way make sense?
- Do the `is-enabled` and `get-variant` input names make sense?

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+[*]
+charset = utf-8
+end_of_line = lf
+indent_style = space
+indent_size = 4
+insert_final_newline = true
+trim_trailing_whitespace = true
+
+[*.md]
+trim_trailing_whitespace = false
```
Comment (sighphyre) [inline .editorconfig:13]: More curious than anything but is there ever a reason to have this as false? I work with Python where not having this on is a compile error, so I turned it on 8 years ago and never looked back

File: action.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  app-name:
+    description: 'The application name for the GitHub Action as you want it to appear in Unleash metrics'
+    type: 'string'
+    required: true
+  api-key:
+    description: 'The frontend API token to use with Unleash'
+    type: 'string'
+    required: true
+  url:
+    description: 'Url to Unleash proxy or frontend api'
```
Comment (sighphyre) [inline action.yml]: It's 2023, let's recommend stuff that's up to date!
```suggestion
    description: 'URL to Unleash Edge or frontend API'
```

File: action.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    type: 'string'
+    required: true
+  environment:
```
Comment (sighphyre) [inline action.yml:16]: This doesn't come from the API key?

File: src/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import { fetch } from './fetch.cjs';
+
+interface MetricsOptions {
```
Comment (sighphyre) [inline src/metrics.ts:3]: Do we need anything in this file? This looks fairly similar to Unleash domain logic, can we not do the job with just using the Proxy Client?

File: src/unleash-action.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+const createClient = (options: ICreateUnleashActionOptions): UnleashClient => {
+    return new UnleashClient({
+        appName: options.appName,
+        url: options.url,
+        clientKey: options.clientKey,
+        environment: options.environment,
+        refreshInterval: 0,
+        metricsInterval: 0,
+        disableMetrics: true,
```
Comment (sighphyre) [inline src/unleash-action.ts:47]: Built in metrics don't work?

File: src/unleash-action.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    private async checkVariants(): Promise<void> {
+        this.variants.forEach((featureName) => {
+            const variant = this.unleash.getVariant(featureName);
+            if (variant.name) {
+                this.metrics.countVariant(featureName, variant.name);
+            }
+            this.metrics.count(featureName, variant.enabled);
+            this.setResult(featureName, variant.enabled);
```
Comment (sighphyre) [inline src/unleash-action.ts:105]: Is it possible to distinguish between a variant for a toggle being on and a toggle being on? These both look like they call setResult with the same values

**Review by sighphyre — APPROVED** [review]
Looks early days and in progress so I'm going to hit the LGTM button!

File: src/unleash-action.ts (hunk 6 above)
Comment (daveleek) [reply src/unleash-action.ts:105]: No, not right now. Do we want to?

File: src/unleash-action.ts (hunk 5 above)
Comment (daveleek) [reply src/unleash-action.ts:47]: It's an async scheduled job over which we have no control and can't await - we can't access the internal metrics either to force sending metrics synchronously. Option B I guess is to expand on what's built into Proxy SDK JS - but that doesn't feel 100% right either

File: src/metrics.ts (hunk 4 above)
Comment (daveleek) [reply src/metrics.ts:3]: I'll take another stab at importing this - but was hard to get at while working on it

File: action.yml (hunk 3 above)
Comment (daveleek) [reply action.yml:16]: Good point! thanks!

File: action.yml (hunk 2 above)
Comment (daveleek) [reply action.yml]: Indeed

File: .editorconfig (hunk 1 above)
Comment (daveleek) [reply .editorconfig:13]: It's been taken from Unleash repo for consistency

**Pull request merged**